### PR TITLE
doc: west: Document build system integration

### PR DESCRIPTION
**Pull request by carlescufi**
Document the build system integration with west and any projects managed
by it.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

### CODE REVIEW

**Comment by carlescufi** [comment]
I drafted the conclusions from today's phone call.
Outstanding items:

1. Should this really be documented in `doc/tools/west` or should it be in `doc/application/` somewhere, where we document the build system?
2. I renamed `zephyr-module.yml` to `zephyr/project.yml`, thoughts?

File: doc/tools/west/repo-tool.rst [hunk 1]
```diff
@@ -327,5 +327,47 @@ These commands perform miscellaneous functions.
...
+  by using the CMake ``add_subdirectory`` command
+- If a matching :file:`Kconfig` is located, it will instruct Kconfig to source
+  it (i.e. process it)
+
+The way that the script determines if a matching :file:`CMakeLists.txt` and
+:file:`Kconfig` are present is the following (in order of precedence):
+
+- If the project contains a file named :file:`<path>/zephyr/project.yml` then
+  its contents are parsed and the following fields are processed if present:
+  - ``cmake:`` points to a folder containing the :file:`CMakeLists.txt` that
```
Comment (dbkinder) [inline doc/tools/west/repo-tool.rst]: Need a blank line before the first bullet list item (in this nested list)

File: doc/tools/west/repo-tool.rst [hunk 2]
```diff
@@ -327,5 +327,47 @@ These commands perform miscellaneous functions.
   that fetches upstream data is run (this behavior can be
   suppressed for the duration of a single command by passing ``--no-update``).
 
+.. _west-mr-build:
+
+Build system integration
+************************
+
+Some of the projects managed by west require to be integrated with the Zephyr
```
Comment (dbkinder) [inline doc/tools/west/repo-tool.rst]:
```suggestion
Some of the projects managed by west must be integrated with the Zephyr
```

File: doc/tools/west/repo-tool.rst [hunk 3]
```diff
@@ -327,5 +327,47 @@ These commands perform miscellaneous functions.
...
+build system, which is based around Kconfig and CMake (see
+:ref:`application` for more information about Zephyr's build system).
+
+The way that the build system discovers and includes west-managed projects when
+building an application is based around the ``west list`` command. The main
+CMake script in zephyr (the :file:`CMakeLists.txt` located at the zephyr
```
Comment (dbkinder) [inline doc/tools/west/repo-tool.rst]: Zephyr should be capitalized when it's a reference to the project or OS (and not a directory name).

File: doc/tools/west/repo-tool.rst [hunk 4]
```diff
@@ -327,5 +327,47 @@ These commands perform miscellaneous functions.
...
+repository root folder) invokes ``west list`` in order to obtain a list of
```
Comment (dbkinder) [inline doc/tools/west/repo-tool.rst]:
```suggestion
repository root folder) invokes ``west list`` to obtain a list of
```

File: doc/tools/west/repo-tool.rst [hunk 5]
```diff
@@ -327,5 +327,47 @@ These commands perform miscellaneous functions.
...
+projects managed by west.
+
+The CMake code in :file:`CMakeLists.txt` retrieves the following
+information for each project using ``west list``:
+- name: The name of the project as specified in the manifest file
```
Comment (dbkinder) [inline doc/tools/west/repo-tool.rst]: need a blank line before the first item of a bullet list

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: doc/tools/west/repo-tool.rst (hunk 3 above)
Comment (mbolivar) [reply doc/tools/west/repo-tool.rst]: In this case it is referring to the "zephyr repository root folder", so it is a directory name.

File: doc/tools/west/repo-tool.rst [hunk 6]
```diff
@@ -327,5 +327,47 @@ These commands perform miscellaneous functions.
...
   suppressed for the duration of a single command by passing ``--no-update``).
 
+.. _west-mr-build:
+
+Build system integration
+************************
+
+Some of the projects managed by west require to be integrated with the Zephyr
+build system, which is based around Kconfig and CMake (see
+:ref:`application` for more information about Zephyr's build system).
```
Comment (mbolivar) [inline doc/tools/west/repo-tool.rst]: Suggest adding:

Examples include vendor HALs and other firmware support libraries.

File: doc/tools/west/repo-tool.rst [hunk 7]
```diff
@@ -327,5 +327,47 @@ These commands perform miscellaneous functions.
...
+
+The CMake code in :file:`CMakeLists.txt` retrieves the following
+information for each project using ``west list``:
+- name: The name of the project as specified in the manifest file
+- path: The path of the project within the west installation
+
+Once it has collected the list, the CMake script performs the following
+operations on each project:
+
+- If a matching :file:`CMakeLists.txt` is located, it will process it directly
```
Comment (mbolivar) [inline doc/tools/west/repo-tool.rst]: `zephyr/CMakeLists.txt` and `zephyr/Kconfig`, right? Or did you change this decision after I left the meeting?

Comment (mbolivar) [reply doc/tools/west/repo-tool.rst]: Also, this seems a bit redundant with the below paragraph that says the same thing.

**Review by mbolivar — CHANGES_REQUESTED** [review]

File: doc/tools/west/repo-tool.rst [hunk 8]
```diff
@@ -327,5 +327,47 @@ These commands perform miscellaneous functions.
...
+build system, which is based around Kconfig and CMake (see
+:ref:`application` for more information about Zephyr's build system).
+
+The way that the build system discovers and includes west-managed projects when
+building an application is based around the ``west list`` command. The main
+CMake script in zephyr (the :file:`CMakeLists.txt` located at the zephyr
+repository root folder) invokes ``west list`` in order to obtain a list of
+projects managed by west.
+
+The CMake code in :file:`CMakeLists.txt` retrieves the following
```
Comment (tejlmand) [inline doc/tools/west/repo-tool.rst]: Just a small question, would you also had written `The C code in file.c` or `The Python code in file.py` ?
If not, I would suggest just: ```The code in :file: `CMakeLists.txt` ```

File: doc/tools/west/repo-tool.rst [hunk 9]
```diff
@@ -327,5 +327,47 @@ These commands perform miscellaneous functions.
...
+
+- If a matching :file:`CMakeLists.txt` is located, it will process it directly
+  by using the CMake ``add_subdirectory`` command
+- If a matching :file:`Kconfig` is located, it will instruct Kconfig to source
+  it (i.e. process it)
+
+The way that the script determines if a matching :file:`CMakeLists.txt` and
+:file:`Kconfig` are present is the following (in order of precedence):
+
+- If the project contains a file named :file:`<path>/zephyr/project.yml` then
```
Comment (tejlmand) [inline doc/tools/west/repo-tool.rst]: as stated in https://github.com/zephyrproject-rtos/zephyr/pull/12938#issuecomment-459638161 , I prefer `zephyr/module.yml`

**Review by tejlmand — COMMENTED** [review]
General good looking.
A couple of small question / comments.

File: doc/tools/west/repo-tool.rst (hunk 7 above)
Comment (carlescufi) [reply doc/tools/west/repo-tool.rst]: I will rework this because it is clearly confusing

File: doc/tools/west/repo-tool.rst (hunk 9 above)
Comment (carlescufi) [reply doc/tools/west/repo-tool.rst]: I'll add a comment below to discuss the names

**Comment by carlescufi** [comment]
@mbolivar @nashif @tejlmand @SebastianBoe 
Can we settle on the names?

1. Per-project yaml file:
  - `<tinycbor>/zephyr/module.yml` **or**
  - `<tinycbor>/zephyr/project.yml` **or**
  - `<tinycbor>/zephyr-module.yml`

2. Default CMake and Kconfig files:
  - `<tinycbor>/zephyr/CMakeLists.txt`
  - `<tinycbor>/zephyr/Kconfig`

**Comment by carlescufi** [comment]
I've reworked this now and addressed everybody's comments

**Comment by tejlmand** [comment]
> Can someone remind what the problem was with having
> 
> ./CMakeLists.txt
> ./Kconfig
> 
> be the default, and allow ./zephyr/*.yml to override this default?
> 
> I wonder if ./CMakeLists.txt being the default was discarded before we introduced the .yml file, so although it didn't make sense at the beginning, it now does. Either that or I am forgetting something.

The problem with that, is that projects unrelated to Zephyr, could very likely be having a `CMakeLists.txt` file at it's root.
And suddenly those projects are required to add `zephyr/[module|project].yml` to avoid that file being parsed (but those projects has nothing to do with Zephyr).

I imagine that it will be very like that users of Zephyr, might created a manifest file as:
```
projects:
  -name: zephyr
  -name: tinycbor
  -name: other-non-zephyr-related-repo
```
and then tell people, if you are gonna work on project A, then just do `west init <project-A>` and you'll get what you need for working (Everything including Zephyr project).

File: doc/tools/west/repo-tool.rst (hunk 9 above)
Comment (tejlmand) [reply doc/tools/west/repo-tool.rst]: Great. I only placed this comment here, so that it would be easy to find the location of `zephyr/project.yml` later.

**Comment by mbolivar** [comment]
> Per-project yaml file:
> a. <tinycbor>/zephyr/module.yml or
> b. <tinycbor>/zephyr/project.yml or
> c. <tinycbor>/zephyr/zephyr.yml or
> d. <tinycbor>/zephyr-module.yml

I hate to split the vote further, but my preference is b, because it matches the name of the manifest file section where these repositories are configured. That makes it a very clear hint to users.

> Default CMake and Kconfig files:
> <tinycbor>/zephyr/CMakeLists.txt
> <tinycbor>/zephyr/Kconfig

+1

**Comment by carlescufi** [comment]
>I hate to split the vote further, but my preference is b, because it matches the name of the manifest file section where these repositories are configured. That makes it a very clear hint to users.

@mbolivar 

I see an issue with that though. In Topology 2 (nrf as manifest repo for example) then nrf is **not** a project in west terms but it may need a `.yml` file in `zephyr/`

**Comment by mbolivar** [comment]
> I see an issue with that though. In Topology 2 (nrf as manifest repo for example) then nrf is **not** a project in west terms but it may need a `.yml` file in `zephyr/`

That is a very good point.

> EDIT: Maybe we need a new term for those repos that get included in the build. `module`? `dependency`? `external repo`? `external project`?

I think module is fine. I change my vote to a.

**Comment by carlescufi** [comment]
@mbolivar @dbkinder @tejlmand pushed a new version, moved the text to the Application Development chapter

**Comment by tejlmand** [comment]
> EDIT: Maybe we need a new term for those repos that get included in the build. module? dependency? external repo? external project?

Good question.
Let me try to put my view.
- `external repo` / `external project` - external to what ? The current github repo, where the manifest is ?
  In that case, when Zephyr uses multiple repos under same `https://github.com/zephyrproject-rtos/<repo>`, i'm not sure that is external enough compared to a company having a manifest as: 
```
remotes:
  - name: company
    url-base: https://<company-internal.somewhere>/
  - name: zephyr
    url-base: https://github.com/zephyrproject-rtos/

projects:
   -  name: internal-project
       remote: company
   - name: zephyr
      remote: zephyr
```
in this example, using `external:`, `external-projects:`, `external-repo:` are all kind of misleading, compared to `projects:`.

- `modules`. This implies that only modules (to Zephyr ?) goes in this list.
But it might be, as also discussed earlier, and described here: https://github.com/zephyrproject-rtos/zephyr/pull/12938#issuecomment-459719359 that a manifest file could contain additional projects (which are not Zephyr related), to allow an easy internal distribution model of multiple repos for development. Additional project may / may not be Zephyr related.

- `projects` This is a list of projects, the list of projects may be a Zephyr module, and thus contain `zephyr/module.yml` or `zephyr/CMakeLists.txt` + `zephyr/Kconfig`, but they don't have to be a Zephyr module.
They can be any kind of project.

Thus, for `west.yml` I still think that a list of `projects:` is the best term, as it contains no indication of whether a project is internal/external/module.

**Comment by carlescufi** [comment]
@tejlmand @mbolivar @dbkinder @nashif 	this should be ready to go now

File: doc/application/application.rst [hunk 10]
```diff
@@ -641,6 +641,73 @@ Zephyr binary into your application directory.
...
+- Debugger integration
+- Silicon Vendor Hardware Abstraction Layers (HALs)
+- Cryptography libraries
+- Filesystems
+- Inter-Process Communication (IPC)
+
+These modules (external projects) must be integrated with the Zephyr
+build system, which is based around Kconfig and CMake (see
+:ref:`application` for more information about Zephyr's build system).
+
```
Comment (carlescufi) [inline doc/application/application.rst]: Clarify module vs project:

- zephyr can never be a module
- a manifest project can be a module
...

Comment (mbolivar) [reply doc/application/application.rst]: > Clarify module vs project:

will wait for this then re-review

**Comment by carlescufi** [comment]
Closing in favor of https://github.com/zephyrproject-rtos/zephyr/pull/13067

**Pull request closed without merging**